src/api/artifact/controller_test.go

```diff
@@ -0,0 +1,303 @@
+	c.assert.Equal(art.ID, artifact.ID)
+	c.assert.Contains(artifact.Tags, &Tag{Tag: *tg})
+	// TODO check other fields of option
+}
+
+func (c *controllerTestSuite) TestAbstract() {
+	// TODO add test case
+}
+
+func (c *controllerTestSuite) TestEnsureArtifact() {
+	repository := &models.RepoRecord{
+		ProjectID:    1,
+		RepositoryID: 1,
+	}
+	digest := "sha256:418fb88ec412e340cdbef913b8ca1bbe8f9e8dc705f9617414c1f2c8db980180"
+
+	// the artifact already exists
+	c.artMgr.On("List").Return(1, []*artifact.Artifact{
+		{
+			ID: 1,
+		},
+	}, nil)
+	art, err := c.ctl.ensureArtifact(nil, repository, digest)
+	c.require.Nil(err)
+	c.artMgr.AssertExpectations(c.T())
+	c.require.NotNil(art)
+	c.Equal(int64(1), art.ID)
+
+	// reset the mock
+	c.SetupTest()
+
+	// the artifact doesn't exist
+	c.artMgr.On("List").Return(1, []*artifact.Artifact{}, nil)
+	c.artMgr.On("Create").Return(1, nil)
+	art, err = c.ctl.ensureArtifact(nil, repository, digest)
+	c.require.Nil(err)
+	c.artMgr.AssertExpectations(c.T())
+	c.require.NotNil(art)
+	c.Equal(int64(1), art.ID)
+}
+
+func (c *controllerTestSuite) TestEnsureTag() {
+	art := &artifact.Artifact{
+		ID:           1,
+		RepositoryID: 1,
+	}
+	tg := "latest"
+
+	// the tag already exists under the repository and is attached to the artifact
+	c.tagMgr.On("List").Return(1, []*tag.Tag{
+		{
+			ID:           1,
+			RepositoryID: 1,
+			ArtifactID:   1,
+			Name:         "latest",
+		},
+	}, nil)
+	err := c.ctl.ensureTag(nil, art, tg)
+	c.require.Nil(err)
+	c.tagMgr.AssertExpectations(c.T())
+
+	// reset the mock
+	c.SetupTest()
+
+	// the tag exists under the repository, but it is attached to other artifact
+	c.tagMgr.On("List").Return(1, []*tag.Tag{
+		{
+			ID:           1,
+			RepositoryID: 1,
+			ArtifactID:   2,
+			Name:         "latest",
+		},
+	}, nil)
+	c.tagMgr.On("Update").Return(nil)
+	err = c.ctl.ensureTag(nil, art, tg)
+	c.require.Nil(err)
+	c.tagMgr.AssertExpectations(c.T())
+
+	// reset the mock
+	c.SetupTest()
+
+	// the tag doesn't exist under the repository, create it
+	c.tagMgr.On("List").Return(1, []*tag.Tag{}, nil)
+	c.tagMgr.On("Create").Return(1, nil)
+	err = c.ctl.ensureTag(nil, art, tg)
+	c.require.Nil(err)
+	c.tagMgr.AssertExpectations(c.T())
+}
+
+func (c *controllerTestSuite) TestEnsure() {
+	repository := &models.RepoRecord{
+		ProjectID:    1,
+		RepositoryID: 1,
+	}
+	digest := "sha256:418fb88ec412e340cdbef913b8ca1bbe8f9e8dc705f9617414c1f2c8db980180"
+	tg := "latest"
+
+	// both the artifact and the tag don't exist
+	c.artMgr.On("List").Return(1, []*artifact.Artifact{}, nil)
+	c.artMgr.On("Create").Return(1, nil)
+	c.tagMgr.On("List").Return(1, []*tag.Tag{}, nil)
+	c.tagMgr.On("Create").Return(1, nil)
+	err := c.ctl.Ensure(nil, repository, digest, tg)
+	c.require.Nil(err)
+	c.artMgr.AssertExpectations(c.T())
+	c.tagMgr.AssertExpectations(c.T())
+}
+
+func (c *controllerTestSuite) TestList() {
+	query := &q.Query{}
+	option := &Option{
+		WithTag:        true,
+		WithScanResult: true,
```

Comment:
scan result and signature are not checked.

Reply:
Scan result and signature isn't appended in controller yet, will test them after implemented